server/routerlicious/packages/routerlicious-base/src/test/nexus/io.spec.ts

```diff
@@ -402,130 +414,447 @@ describe("Routerlicious", () => {
-										signal.content,
-										"Signal content mismatch",
-									);
-									assert.equal(
-										firstConnectMessage.clientId,
-										signal.clientId,
-										"Client ID mismatch",
+					function verifyExpectedClientSignals(
+						clients: TestSignalClient[],
+						expectedSignals: ISignalMessage[],
+					) {
+						clients.forEach((client, index) => {
+							assert.equal(
+								client.signalsReceived.length,
+								expectedSignals.length,
+								`User ${index + 1} should have received ${
+									expectedSignals.length
+								} signal(s)`,
+							);
+							expectedSignals.forEach((signal, signalIndex) => {
+								const receivedSignal = client.signalsReceived[signalIndex];
+								Object.keys(signal).forEach((property) => {
+									assert.deepEqual(
+										receivedSignal[property],
+										signal[property],
+										`User ${index + 1} signal ${
+											signalIndex + 1
+										} ${property} mismatch`,
 									);
-									resolve();
 								});
-							}),
-							new Promise<void>((resolve) => {
-								secondSocket.on("signal", (signal: ISignalMessage) => {
-									assert.equal(
+							});
+						});
+					}
+
+					function sendAndReturnExpectedSignal(
+						client: TestSignalClient,
+						signal: unknown,
+					): ISignalMessage {
+						client.socket.send("submitSignal", client.clientId, [signal]);
+						let expectedSignalMessage: ISignalMessage;
+						if (client.version === 2 && isSentSignalMessage(signal)) {
+							expectedSignalMessage = {
+								...signal,
+								clientId: client.clientId,
+							};
+						} else {
+							expectedSignalMessage = {
+								clientId: client.clientId,
+								content: signal,
+							};
+						}
+						return expectedSignalMessage;
+					}
+
+					function listenForNacks(client: TestSignalClient) {
+						client.socket.on("nack", (reason: string, nackMessages: INack[]) => {
+							client.nacksReceived.push(nackMessages[0]);
+						});
+					}
+
+					function checkNack(
+						client: TestSignalClient,
+						expectedNackMessageContent: string,
+					) {
+						assert.equal(
+							client.nacksReceived.length,
+							1,
+							"Client should have received 1 nack",
+						);
+						const nackMessage = client.nacksReceived[0];
+						assert.equal(nackMessage.content.code, 400, "Nack code should be 400");
+						assert.equal(
+							nackMessage.content.type,
+							NackErrorType.BadRequestError,
+							"Nack type should be BadRequestError",
+						);
+						assert.deepEqual(
+							nackMessage.content.message,
+							expectedNackMessageContent,
+							"Nack message should be 'Invalid signal message'",
+						);
+					}
+
+					const stringSignalContent = "TestSignal";
+
+					let clients: TestSignalClient[];
+
+					const numberOfClients = 3; // Change the amount of clients to test with (at least 2 required)
+
+					assert(numberOfClients > 1, "Test requires at least 2 clients");
+
+					describe("with v1 clients", () => {
```

Comment:
All of the tests in here apply to every mix of clients, but this only tests all of these with v1 only. Some of these are tested in later sections, but not all.
We can use a slightly more advanced loop to have the same tests run for multiple scenarios. I will write up the pattern and reply.

Reply:
You can replace that describe through the "sending one signal" with this:
```typescript
					[
						["with v1 clients", () => 1 as const] as const,
						["with v2 clients", () => 2 as const] as const,
						[
							"with v1 and v2 clients",
							(index: number) =>
								(1 + (index % 2)) as 1 | 2,
						] as const,
					].forEach(([description, fnVersion]) =>
						describe(description, () => {
							const clientVersion: TestSignalClient["version"][] = [];
							for (let i = 0; i < numberOfClients; i++) {
								clientVersion.push(fnVersion(i));
							}
							beforeEach(async () => {
								clients = await Promise.all(
									clientVersion.map((version) =>
										createClient(testId, testTenantId, testSecret, version),
									),
								);
								listenForSignals(clients);
							});
							describe("sending one signal", () => {
								[0, 1].forEach((clientIndex) =>
									it(`from client ${clientIndex} (v${clientVersion[clientIndex]}) should broadcast signal to all connected clients`, () => {
										const expectedSignal = sendValidAndReturnExpectedSignal(
											clients[0],
											stringSignalContent,
										);

										verifyExpectedClientSignals(clients, [expectedSignal]);
									}),
								);
```

To use that exact code there are a few other things you'd need.
1. This helper that know how to send a basic valid signal:
```typescript
					function sendValidAndReturnExpectedSignal(
						client: TestSignalClient,
						content: unknown,
					): ISignalMessage {
						const signal = client.version === 2 ? { content } : content;
						return sendAndReturnExpectedSignal(client, signal);
					}
```
2. I changed `createClient` to always take a version (1 | 2) instead of a boolean.

Optionally you can move the other `describe("with v2 clients")` and `describe("with v1 and v2 clients")` under that loop describe. You can say
```typescript
if (description === "with v2 clients") {
   ... the other cases you had
} else if (description === "with v1 and v2 clients") {
   ...
```
the advantages are 1) that beforeEach is handled and 2) (just nice) that mocha will group them all together versus having two "with v2 clients" and two "with v1 and v2 clients" groups.

Reply:
I'm struggling to parse what this part is doing

```typescript 

describe("sending one signal", () => {
        [0, 1].forEach((clientIndex) =>
	        it(`from client ${clientIndex} (v${clientVersion[clientIndex]}) should broadcast signal to all connected clients`, () => {
		        const expectedSignal = sendValidAndReturnExpectedSignal(
			        clients[0],
			        stringSignalContent,
		        );
        
		        verifyExpectedClientSignals(clients, [expectedSignal]);
	        }),
        );
```

Reply:
I updated it to follow this pattern. I interpreted the `clientIndex` part a little different then as showed, I used it to index the clients array.